70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+				IWindow::SCreationParams params{
+					.callback = core::make_smart_refctd_ptr<ISimpleManagedSurface::ICallback>(),
+					.x = 32,
+					.y = 32,
+					.width = WIN_WIDTH,
+					.height = WIN_HEIGHT,
+					.flags = IWindow::ECF_HIDDEN | IWindow::ECF_BORDERLESS | IWindow::ECF_RESIZABLE,
+					.windowCaption = "FLIPFluidsApp"
+				};
+				params.callback = windowCallback;
+				const_cast<std::remove_const_t<decltype(m_window)>&>(m_window) = m_winMgr->createWindow(std::move(params));
+			}
+
+			auto surface = CSurfaceVulkanWin32::create(smart_refctd_ptr(m_api), smart_refctd_ptr_static_cast<IWindowWin32>(m_window));
+			const_cast<std::remove_const_t<decltype(m_surface)>&>(m_surface) = nbl::video::CSimpleResizeSurface<CSwapchainFramebuffersAndDepth>::create(std::move(surface));
+		}
+
+		if (m_surface)
+			return { { m_surface->getSurface() } };
+
+		return {};
+	}
+
+	inline bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		m_inputSystem = make_smart_refctd_ptr<InputSystem>(logger_opt_smart_ptr(smart_refctd_ptr(m_logger)));
+
+		if (!device_base_t::onAppInitialized(std::move(system)))
+			return false;
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+			return false;
+
+		// init grid params
+		usePreset(CENTER_DROP);
+		
+		WorkgroupCountParticles = (numParticles + WorkgroupSize - 1) / WorkgroupSize;
+		WorkgroupCountGrid = (numGridCells + WorkgroupSize - 1) / WorkgroupSize;
+
+		{
+			float zNear = 0.1f, zFar = 10000.f;
+			core::vectorSIMDf cameraPosition(14, 8, 12);
+			core::vectorSIMDf cameraTarget(0, 0, 0);
+			matrix4SIMD projectionMatrix = matrix4SIMD::buildProjectionMatrixPerspectiveFovLH(core::radians(60.0f), float(WIN_WIDTH) / WIN_HEIGHT, zNear, zFar);
+			camera = Camera(cameraPosition, cameraTarget, projectionMatrix, 1.069f, 0.4f);
+
+			m_pRenderParams.zNear = zNear;
+			m_pRenderParams.zFar = zFar;
+		}
+		m_pRenderParams.radius = m_gridData.gridCellSize * 0.4f;
+
+		// create buffers
+		video::IGPUBuffer::SCreationParams params = {};
+		params.size = sizeof(SGridData);
+		params.usage = IGPUBuffer::EUF_UNIFORM_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT | IGPUBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF;
+		createBuffer(gridDataBuffer, params);
+
+		params.size = 2 * sizeof(float32_t4);
+		createBuffer(pressureParamsBuffer, params);
+		
+		params.size = sizeof(SMVPParams);
+		createBuffer(cameraBuffer, params);
+
+		params.size = sizeof(SParticleRenderParams);
+		createBuffer(pParamsBuffer, params);
+
+		params.size = numParticles * sizeof(Particle);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_SRC_BIT;
+		createBuffer(particleBuffer, params);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+		createBuffer(tempParticleBuffer, params);
+
+		params.size = numGridCells * sizeof(uint32_t2);
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+		createBuffer(gridParticleIDBuffer, params);
```

Comment:
after you do Global Counting sort, you can rename this to particleHistogramImage